### PR TITLE
Add: flutter build ios [--[no-]simulator]

### PR DESCRIPTION
**Pull request by chinmaygarde**
This allows the user to build an iOS application bundle from the command line even if no device is attached to the workstation.

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @devoncarew

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    printStatus('Building the application for $logTarget.');
+
+    bool result = await buildIOSXcodeProject(app, buildForDevice: !forSimulator);
+
+    if (!result) {
+      printError('Encountered error while building for $logTarget.');
+      return 1;
+    }
+
+    printStatus('Done.');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: Can we print out here where it build? `Built foo/bar/baz.`

File: packages/flutter_tools/lib/src/commands/build.dart [hunk 2]
```diff
@@ -8,12 +8,14 @@ import 'dart:io';
...
 import '../runner/flutter_command.dart';
 import 'build_apk.dart';
+import 'build_ios.dart';
 import 'build_flx.dart';
 
 class BuildCommand extends FlutterCommand {
   BuildCommand() {
     addSubcommand(new BuildApkCommand());
     addSubcommand(new BuildCleanCommand());
+    addSubcommand(new BuildIOSCommand());
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build.dart]: Is this to build ipas? Perhaps call this `ipa` for parity with the `apk` command. Or we could rename both to `android` and `ios`. In any case, there should be some similarity.

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import '../application_package.dart';
+import '../build_configuration.dart';
+import '../globals.dart';
+import '../ios/mac.dart';
+import '../runner/flutter_command.dart';
+
+class BuildIOSCommand extends FlutterCommand {
+  BuildIOSCommand() {
+    argParser.addFlag('simulator', help: 'Build for the iOS simulator instead of the device');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: perhaps suffix with ` (macos ony)`; we have a few other commands that have these suffixes

**Comment by devoncarew** [comment]
lgtm w/ some naming questions

File: packages/flutter_tools/lib/src/commands/build.dart (hunk 2 above)
Comment (chinmaygarde) [inline packages/flutter_tools/lib/src/commands/build.dart]: No, these commands just build the application bundle. Building IPAs is a little bit more involved. When I eventually get around to it though, I can add it as a subcommand to ios.

Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build.dart]: OK, let's keep this as `ios`.

**Comment by chinmaygarde** [comment]
@devoncarew another look?

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 4]
```diff
@@ -114,9 +115,23 @@ Future<bool> buildIOSXcodeProject(ApplicationPackage app, { bool buildForDevice
...
+    'ONLY_ACTIVE_ARCH=YES',
   ];
 
+  if (buildDirectory != null) {
+    if (!buildDirectory.existsSync()) {
+      printError('The specified build directory ${buildDirectory.path} does not exist');
+      return false;
+    }
+
+    commands.add('TARGET_BUILD_DIR=${buildDirectory.absolute.path}');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Will this fail w/ spaces in the path? Perhaps escape the space or quote the path.

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  @override
+  final String name = 'ios';
+
+  @override
+  final String description = 'Build an iOS application bundle (Mac OSX host only).';
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: How about just `(Mac OS only)`

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    await Future.wait([
+      downloadToolchain(),
+      downloadApplicationPackages(),
+    ], eagerError: true);
+
+    IOSApp app = applicationPackages.iOS;
+
+    if (app == null) {
+      printError('Application not configured for iOS');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: `No ios/ directory found; run 'flutter create' to generate an ios/ directory for your app.`?

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    bool forSimulator = argResults['simulator'];
+
+    String logTarget = forSimulator ? "simulator" : "device";
+
+    printStatus('Building the application for $logTarget.');
+
+    Directory buildDir = new Directory(path.join('build', 'ios_$logTarget'));
+
+    if (!buildDir.existsSync())
+      await buildDir.create();
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: We can just use buildDir.createSync() here, w/o an await

**Comment by devoncarew** [comment]
missed the 2nd review, but added some comments -

**Comment by chinmaygarde** [comment]
Addressing. I also found a couple of other issues as I was adding a Travis bot for this. Will send another PR.

**Pull request merged**